Introduce new deserialization API

wprzytula opened the pull request:
# Here it is, finally!

After #1004 with preparation, #970 with lower layer and #1024 with derive macros,
this PR introduces the upper layers of the new deserialization framework.

## The Big Picture

1. Legacy structures are prefixed `Legacy`, deprecated and planned to be removed soon:
   - `QueryResult` -> `LegacyQueryResult`
   - `(Typed)RowIterator` -> `Legacy(Typed)RowIterator`
1. New structures are added in the upper layer that support lazy deserialization.
   - `RawRows` replace `Rows`. They contain deserialized metadata, but rows are still in serialized form.
   - `RawIterator`, `TypedRowIterator` and `TypedRowStream` replace `LegacyRowIterator` and `LegacyTypedRowIterator`.
   - `QueryResult` replaces `LegacyQueryResult`.
1. `RawRowsLendingIterator` is added (see details below)
1. Both `QueryResult`'s and paging iterators' API is changed wrt column specs inspection (see details below).
1. `Session` is made generic over the deserialization API kind, so that both old and new API can be used in the transition period by the users. The choice is done on `SessionBuilder`'s level, by choosing either `build()` or `build_legacy()` method.
1. Internal interfaces (e.g. `Connection`'s API, especially `query_iter()` widely used in `topology.rs`) are adjusted to use the new API.
1. Tests, examples and doctests are adjusted to use the new API.
1. Deserialization logic of collections is changed in a way that now it is possible to deserialize empty collections into `Collection` instead of `Option<Collection>`, even though the DB represents empty collections as nulls.
1. A test is added based on #1007, showing that now it is indeed possible to deserialize empty collections into `Collection` instead of `Option<Collection>`.

## Details

### `RawRowsLendingIterator`

It is added in tandem with the already present (the low level, `scylla_cql` one, not the `iterator.rs` one!) `RowIterator`. The difference is that `RawRowsLendingIterator` owns the frame and the metadata, whereas `RowIterator` borrows it. The lending one is needed in session iterator interface, and the borrowing one is used with `QueryResult`.

### Achieving old behaviour for dynamic deserialization purposes

If one still wants to deserialize arbitrary rows (as the old framework would do) and only then try to convert them dynamically to some proper types, this is doable by using the old `Row` struct as the target type, e.g.
```rust
let qr = session.query("SELECT a FROM ks.table")?;
let rows = qr.rows::<Row>()?;
for row in rows {
    // Type check and deserialization succeeds for any valid CQL type.

    // Do arbitrary dynamic magic with `row`.
}
```

### `ColumnSpecs`

The new `ColumnSpecs` struct is introduced that provides view over columns specification, but exposes only `TableSpecView` and `ColumnSpecView` structs instead of `TableSpec` and `ColumnSpec`, allowing us to modify the latter in the future without introducing breaking changes. One planned change is moving `table_spec` out of `column_spec` for reduced memory footprint (see #956), but as CQL protocol allows different table specs for different columns and we can't predict whether joins won't be introduced at some point in the future, we retain this flexibility by imposing a new layer of abstraction (the views).

### New paging iterators

Instead of `RowIterator` (returning `Row`s) and `TypedRowIterator` (returning instances of the target type) both implementing `Stream`, now we have the following:
- `RawIterator`
  - cannot implement `Stream`, because returns `ColumnIterator`s that borrow from it,
  - provide `type_check()` and `next()` methods that can be used for low-level, manual deserialization (not recommended for ordinary users)
  - supports deserializing manually borrowed types (such as `&str`).
- `TypedRowIterator`
  - created by calling `into_typed::<TargetType>()` on `RawIterator`,
  - type checks upon creation,
  - supports deserializing automatically borrowed types (such as `&str`),
  - does not implement `Stream` in order to support borrowed types.
- `TypedRowStream`
  - created by calling `into_stream()` on `TypedRowIterator`,
  - implements `Stream` and hence does not support borrowed types.

## TODO

- Documentation needs to be adjusted. This is to be done in a follow-up.
- A migration guide is planned for a follow-up.
- Deprecation notices for the legacy API are done, but delayed for a follow-up not to make this PR too huge.

Fixes: #964
Fixes: #1001
_Nearly_resolves_: #462

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [ ] I have adjusted the documentation in `./docs/source/`.
- [x] I added appropriate `Fixes:` annotations to PR description.

wprzytula commented:
v1.0.1: clippy fixes

wprzytula commented:
v1.0.2: docstrings fixes

wprzytula commented:
v1.0.3: tests fixes.

wprzytula commented:
v1.0.4: docstring fixes.

Lorak-mmk reviewed (changes requested):
Posting the comments that I already made so that they are not lost

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -459,6 +460,72 @@ impl Row {
     }
 }
 
+/// Rows response, in partially serialized form.
+// TODO: We could provide ResultMetadata in a similar, lazily
+// deserialized form - now it can be a source of allocations
+#[derive(Debug, Default)]
+pub struct RawRows {
+    metadata: ResultMetadata,
```
Absolutely, this is what I've done experimentally after posting the PR and it works. Will introduce it in the new smaller PRs.

wprzytula commented on scylla/src/transport/query_result.rs:
```diff
@@ -0,0 +1,719 @@
+pub struct TableSpecView<'res> {
+    table_name: &'res str,
+    ks_name: &'res str,
+}
```
As `TableSpec` is defined in `scylla-cql`, it must have public API (a public constructor and/or public fields), for `scylla` crate to be able to access it. Hence, we don't want to expose `TableSpec` to users at all; instead, it's better to only expose types defined in `scylla`, because then we can truly control visibility using `pub` and `pub(crate)`. 

This is a pattern that I believe we should employ commonly, because then we will be able to make code adjustments to inner types, even changing their names and/or types, without introducing breaking API changes.

Lorak-mmk replied:
Is a public contructor a problem in this case? I don't see why, but probably I don't understand something.

This patterns means basically duplicating a lot of code from scylla-cql. If we have to do this then maybe scylla-cql should not exist?

wprzytula replied:
1. scylla-cql was created specifically to allow code reuse for scylla-proxy. As long as scylla-proxy exists and we don't want to duplicate low-level ser/de code there, we need scylla-cql.
2. You see this as duplicating code. I see this as introducing a new layer - stable abstraction. This is something that a stable driver should have, independently of its inner working. I have an impression that Rust driver has not yet developed such layer, and its API more-or-less reflects its inner implementation details. For the sake of future compatibility and more freedom to modify things after 1.0, I believe we need such layer, consisting of exposed yet extremely stable types. Even if in some cases it boils down to an identity-or-nearly-identity layer.

wprzytula replied:
To be honest, I hate the dual world of scylla-cql and scylla. I feel unsafe about those `pub`s in scylla-cql.
Perhaps we could at least move some part of scylla-cql to scylla. It seems that scylla-proxy does not use the serialization nor deserialization framework anyway (even though I initially intended it to do so - to be capable of e.g. injecting arbitrary data).

wprzytula commented on scylla/src/transport/query_result.rs:
```diff
@@ -0,0 +1,719 @@
+    /// Transforms itself into the legacy result type, by eagerly deserializing rows
+    /// into the Row type. This is inefficient, and should only be used during transition
+    /// period to the new API.
+    pub fn into_legacy_result(self) -> Result<LegacyQueryResult, ParseError> {
+        if let Some(raw_rows) = self.raw_rows {
+            let deserialized_rows = raw_rows
+                .rows_iter::<Row>()?
+                .collect::<Result<Vec<_>, DeserializationError>>()?;
+            let serialized_size = raw_rows.rows_size();
+            let metadata = raw_rows.into_metadata();
+            Ok(LegacyQueryResult {
```
It can fail in a lot of ways, just because the serialized data can be malformed. In such case, DeserializationError will be returned.
I don't see how this is a complication. Question marks can be added easily.

wprzytula commented on scylla/src/transport/query_result.rs:
```diff
@@ -0,0 +1,719 @@
+        fn serialize_cells(cells: impl IntoIterator<Item = Option<impl AsRef<[u8]>>>) -> Bytes {
+            let mut bytes = BytesMut::new();
+            for cell in cells {
+                types::write_bytes_opt(cell, &mut bytes).unwrap();
```
I can be, of course, separate. My idea was that those subtests are so small that it's not worth introducing overhead for that many that small tests.

Lorak-mmk replied:
I don't think there is any measurable overhead, and until there is proof otherwise we should value readability over such optimizations

wprzytula replied:
Having read the test's code once again, I think it's more meaningful and compact as one. I don't see a good (i.e. bringing more sense) way to split it.

wprzytula commented on scylla/src/transport/query_result.rs:
```diff
@@ -0,0 +1,719 @@
+    }
+}
+
+/// A view over specification of a column returned by the database.
+#[derive(Debug, Clone, Copy)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub struct ColumnSpecView<'res> {
+    table_spec: TableSpecView<'res>,
+    name: &'res str,
+    typ: &'res ColumnType,
+}
```
I answered this in the thread related to TableSpecView.

wprzytula commented on scylla/src/transport/query_result.rs:
```diff
@@ -0,0 +1,719 @@
+pub struct ColumnSpecs<'res> {
+    specs: &'res [ColumnSpec],
+}
+
```
This type is intentionally introduced to provide a stable abstraction over possibly unstable inner types.

wprzytula commented on scylla/src/transport/iterator.rs:
```diff
@@ -849,8 +851,8 @@ where
-                    paging_state = rows.metadata.paging_state.take();
+                NonErrorResponse::Result(result::Result::Rows(rows)) => {
+                    paging_state = rows.metadata().paging_state.clone();
                     let (proof, send_result) = self
```
Because `metadata()` returns immutable reference. This is how @piodul designed it. I'm reluctant to do in-place modifications on metadata; OTOH, cloning page state does not involve allocations.

wprzytula replied:
Keep in mind that once paging state is decoupled from `ResultMetadata`, this is no longer going to be needed.

wprzytula commented on scylla/src/transport/iterator.rs:
What can I do about it to make it possible to be reviewed? I just moved some code and this is the result. Can it be done even cleaner?

Lorak-mmk replied:
I didn't mean that it is bad that it's not possible to review it or that you need to fix it. Just wanted to note that I did not review it.

wprzytula commented on scylla/src/transport/iterator.rs:
```diff
@@ -898,17 +965,147 @@ impl LegacyRowIterator {
+
+impl<RowT> Unpin for TypedRowStream<RowT> {}
+
+/// Stream implementation for TypedRowStream.
+///
+/// It only works with owned types! For example, &str is not supported.
+impl<RowT> Stream for TypedRowStream<RowT>
+where
```
> So a user that uses `execute_iter` won't be able to deserialize to non-owned types?

Fortunately, no!

The @piodul's version of the refactor indeed had that limitation. The reason is that Stream, analogously to Iterator, does not use GATs and hence does not permit lending from itself. Therefore, if we want to have Stream implemented, we can't return borrowed items.
But then I discovered that we can introduce two separate types:
- `TypedRowIterator` - does not implement Stream, implements `next()` method that permits lending from itself;
- `TypedRowStream` - implements Stream, but has a limitation for owned types only.

TypedRowIterator is convertible to TypedRowStream for free. This solves the problem.

wprzytula commented on scylla-cql/src/frame/frame_errors.rs:
```diff
@@ -44,6 +44,8 @@ pub enum ParseError {
     #[error(transparent)]
     DeserializationError(#[from] DeserializationError),
     #[error(transparent)]
+    DeserializationTypeCheckError(#[from] TypeCheckError),
+    #[error(transparent)]
     IoError(#[from] std::io::Error),
     #[error("type not yet implemented, id: {0}")]
     TypeNotImplemented(u16),
```
> I wonder if it isn't better to add a variant in a commit where it is first used - that way the person reading the commit doesn't have to guess why this variant is needed.

OTOH, this adds clutter to the other commit. And this variant really should have been present from the very beginning.

wprzytula commented:
v2.0:
- rebased on main, incorporating paging-related changes,
- resolved some Karol's comments.

wprzytula commented:
v2.1.1:
- fixed doctests.

wprzytula commented:
v3.0: Rebased on #1093. Removed obsolete parts of the cover letter.

muzarski commented on scylla/src/transport/errors.rs:
```diff
@@ -360,20 +362,28 @@ pub enum SchemaVersionFetchError {
+    TracesSessionInvalidColumnType(TypeCheckError),
+
+    /// Response to system_traces.session failed to deserialize.
+    #[error("Response system_traces.session failed to deserialize: {0}")]
```
Response ***to***

muzarski commented on scylla/src/transport/errors.rs:
```diff
@@ -360,20 +362,28 @@ pub enum SchemaVersionFetchError {
+    TracesEventsInvalidColumnType(TypeCheckError),
+
+    /// Response to system_traces.events failed to deserialize.
+    #[error("Response system_traces.events failed to deserialize: {0}")]
```
ditto: Response ***to***

wprzytula commented:
v3.0.1: introduced fixes suggested by @muzarski.

wprzytula commented:
Added the following TODO to the cover letter:
> A follow-up is going to bring methods for conversions between Sessions of both APIs, so that both façades are cheaply convertible into each other (by sharing their inner data with an Arc). This will enable using both APIs in one application with minimal changes needed.

Lorak-mmk commented on scylla/src/transport/connection.rs:
Commit: "connection: switch to the new deserialization framework "

Commit message says "Adjusts the methods of Connection..." but I only see a single method adjusted - fetch_schema_version

wprzytula replied:
That's correct.
let _A_ := The set of methods on Connection that used old deserialization framework before this commit
let _B_ := `query_iter()`, `execute_iter()`

_A_ - _B_ = _{`fetch_schema_version`}_

wprzytula replied:
Fixed the commit description.

Lorak-mmk commented on scylla/src/transport/caching_session.rs:
Commit: "caching_session: fix docstring references "

If references were broken, then is it true that "All commits compile, pass static checks and pass test."? I guess that rustdoc doesn't fail because of such problems so there is no error?

wprzytula replied:
> did you verify with cargo doc that all those references are rendered and linked correctly? not only in VsCode but more importantly in generated docs.

I did.

Lorak-mmk commented on scylla/src/transport/connection.rs:
```diff
@@ -2479,6 +2478,8 @@ mod tests {
             .query_iter(select_query.clone())
             .await
             .unwrap()
+            .rows_stream::<Row>()
+            .unwrap()
             .try_collect::<Vec<_>>()
             .await
             .unwrap();
```
Why do you use `Row`? It looks like the types are known here.

wprzytula replied:
The old code used untyped API, that's why I resorted to `Row`. I didn't want to invest my time in checking what's the expected type of this row.

Lorak-mmk commented on scylla/src/transport/connection.rs:
```diff
@@ -2517,6 +2519,8 @@ mod tests {
             ))
             .await
             .unwrap()
+            .rows_stream::<Row>()
+            .unwrap()
             .try_collect::<Vec<_>>()
             .await
             .unwrap();
```
ditto

Lorak-mmk commented on scylla/src/transport/topology.rs:
```diff
@@ -799,6 +803,13 @@ async fn query_peers(conn: &Arc<Connection>, connect_port: u16) -> Result<Vec<Pe
     let peers_query_stream = conn
         .clone()
         .query_iter(peers_query)
+        .map(|it| {
+            it.and_then(|it| {
+                it.rows_stream::<NodeInfoRow>()
+                    .map_err(|err| RowsParseError::from(err).into())
+                    .map(|iter| iter.into_stream())
+            })
+        })
         .into_stream()
         .try_flatten()
         .and_then(|row_result| future::ok((NodeInfoSource::Peer, row_result)));
```
Before we had just `.into_stream()`. now there is `into_stream()` x2 + `.rows_stream()`  - why?
There is also a lot of nesting and shadowing - it's quite difficult to read. Please make it more readable. Few ideas:
- Explicit types
- More descriptive variable names
- Intermediate variables and `if let`
- Definitely less nesting and indentation

wprzytula commented on scylla/src/transport/topology.rs:
```diff
@@ -905,11 +922,15 @@ async fn create_peer_from_row(
 
-fn query_filter_keyspace_name<'a>(
+fn query_filter_keyspace_name<'a, R>(
     conn: &Arc<Connection>,
     query_str: &'a str,
     keyspaces_to_fetch: &'a [String],
-) -> impl Stream<Item = Result<Row, QueryError>> + 'a {
+    convert_typecheck_error: impl FnOnce(TypeCheckError) -> MetadataError + 'a,
+) -> impl Stream<Item = Result<R, QueryError>> + 'a
+where
+    R: for<'r> DeserializeRow<'r, 'r> + 'static,
+{
```
I'm not a fan of this idea. For me, the fact that a type is `'static` is completely independent of the `DeserializeRow` implementation, and we shouldn't add another name to mean both of those things. Mentioning this in a straightforward way is the clearest.

wprzytula commented on scylla/src/transport/topology.rs:
```diff
@@ -905,11 +922,15 @@ async fn create_peer_from_row(
 
-fn query_filter_keyspace_name<'a>(
+fn query_filter_keyspace_name<'a, R>(
     conn: &Arc<Connection>,
     query_str: &'a str,
     keyspaces_to_fetch: &'a [String],
-) -> impl Stream<Item = Result<Row, QueryError>> + 'a {
+    convert_typecheck_error: impl FnOnce(TypeCheckError) -> MetadataError + 'a,
+) -> impl Stream<Item = Result<R, QueryError>> + 'a
+where
+    R: for<'r> DeserializeRow<'r, 'r> + 'static,
+{
```
`FnOnce` unfortunately cannot be used with `&dyn` - calling `FnOnce` need ownership of it. I'll investigate using `&dyn Fn` here.

wprzytula replied:
Actually, `impl FnOnce` does not hurt us, because `R` makes this function monomorphised anyway. I made an optimisation, though: I degenerised a large chunk of this function, so that the monomorphisation penalty is smaller.

wprzytula commented on scylla/src/transport/cql_types_test.rs:
```diff
@@ -77,9 +74,9 @@ async fn init_test(table_name: &str, type_name: &str) -> LegacySession {
+    let session: Session = init_test(type_name, type_name).await;
     session.await_schema_agreement().await.unwrap();
 
     for test in tests.iter() {
```
I believe `DeserializeOwned{Row,Value}` is a better fit

wprzytula replied:
I added them in the new commit, put as the last.

wprzytula commented:
v3.1: Addressed all comments.

Lorak-mmk commented on scylla/src/transport/session.rs:
Commit: session: de-genericise internal query/exec functions 

It does not de-genericises do_* methods because they are still generic over values: impl SerializeRow

Please adjust the commit message, or de-genericise them for real.

wprzytula replied:
Adjusted the commit message.
I have an idea how to reduce the monomorphisation cost there, left for another PR in the future.

Lorak-mmk commented on scylla/src/lib.rs:
```diff
@@ -230,6 +230,13 @@ pub mod deserialize {
             UdtIterator, UdtTypeCheckErrorKind,
         };
     }
+
+    // Shorthands for better readability.
+    #[cfg_attr(not(test), allow(unused))]
+    pub(crate) trait DeserializeOwnedValue: for<'r> DeserializeValue<'r, 'r> {}
+    impl<T> DeserializeOwnedValue for T where T: for<'r> DeserializeValue<'r, 'r> {}
+    pub(crate) trait DeserializeOwnedRow: for<'r> DeserializeRow<'r, 'r> {}
+    impl<T> DeserializeOwnedRow for T where T: for<'r> DeserializeRow<'r, 'r> {}
```
Question regarding this bounds (applicable also before this commit): Why use a single lifetime?
In other words, why `for<'r> DeserializeValue<'r, 'r>` and not `for<'frame, 'metadata> DeserializeValue<'frame, 'metadata>`? I'm not saying it is wrong - I just wanted to understand the reasoning.

The second bound is more restrictive, right? Maybe the first one is just enough, or maybe there is no difference?

wprzytula replied:
The first one is just enough.

Meanwhile, the second one is too restrictive, and I believe impossible to be satisfied!
Namely,
```rust
for<'frame, 'metadata> DeserializeValue<'frame, 'metadata>
```
would require that `DeserializeValue` is implemented for all possible ('frame, 'metadata) pairs, including those where `'frame: 'metadata` does not hold true - and this contradicts the said logical lifetime relationship that normal `DeserializeValue` implementors impose.

Lorak-mmk replied:
As proved experimentally, `for<'frame, 'metadata> DeserializeValue<'frame, 'metadata>` works, and those traits could use that.

Given that they are internal and not exposed to user, it does not matter what they use.

Lorak-mmk reviewed (changes requested):
Look quite good, just a few minor comments.

wprzytula commented on examples/basic.rs:
```diff
@@ -53,39 +55,38 @@ async fn main() -> Result<()> {
     let mut iter = session
         .query_iter("SELECT a, b, c FROM examples_ks.basic", &[])
         .await?
-        .into_typed::<(i32, i32, String)>();
+        .rows_stream::<(i32, i32, String)>()?;
     while let Some((a, b, c)) = iter.try_next().await? {
         println!("a, b, c: {}, {}, {}", a, b, c);
     }
```
I fully agree with this remark, but I believe it's something to be done later, not in this PR (and maybe even not necessarily in the next release).

wprzytula replied:
In this particular case, we are using the `Stream`, which does not permit lending (and hence &str).

wprzytula commented:
v3.1.1: amended commit message about de-genericisation - now it mentions _**partial**_ de-genericisation.

muzarski reviewed (approved):
Looks good

wprzytula commented:
v3.2: Added 2 new commits on top of the PR:
- renamed `RowIteratorWorker` to `PagerWorker`,
- made lifetime constraints in `QueryPager::rows_stream()` more rigid to prevent errors.
